File: 144/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def preorderTraversal(self, root):
        if root is None:
            return []
        result = []
        stack = [root]
        while stack:
            visit = stack.pop()
            result.append(visit.val)
            if visit.right:
                stack.append(visit.right)
            if visit.left: 
                stack.append(visit.left)
        return result

File: 144/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
    if root.nil?
        return []
    end
    
    result = []
    stack = [root]
    while !stack.empty?
        visit = stack.pop
        result.push(visit.val)
        if !visit.right.nil?
            stack.push(visit.right)
        end
        if !visit.left.nil?
            stack.push(visit.left)
        end
    end
    return result
end